### ISSUE TITLE
Generated meson.build from Cargo crate references undefined variable

### ISSUE THREAD

**Issue by 2xsaiko** [mesonbuild/meson#13855]
**Describe the bug**
Meson's cargo importer generates a meson.build that is referencing an undefined variable for the crate libflate_lz77, which is a transitive dependency of the crate I'm trying to import.

Might be related to these two crates depending on each other, or dev-dependencies: libflate depends on libflate_lz77 and libflate_lz77 has libflate in its dev-dependencies.

> Executing subproject matfile:libflate-2-rs:libflate_lz77-2-rs method cargo
> 
> libflate_lz77-2-rs| Generated Meson AST: /Users/saiko/Src/test/build/subprojects/libflate_lz77-2.1.0/meson.build
> libflate_lz77-2-rs| Project name: libflate_lz77
> libflate_lz77-2-rs| Project version: 2.1.0
> libflate_lz77-2-rs| Rust compiler for the host machine: rustc -C linker=cc (rustc 1.82.0)
> libflate_lz77-2-rs| Rust linker for the host machine: rustc -C linker=cc ld.lld 19.1.2
> 
> subprojects/libflate_lz77-2.1.0/meson.build:25:-1: ERROR: Unknown variable "libflate_options".

The first couple lines from the meson.build in question (note libflate_options not being defined unlike the others):
```
project(
  'libflate_lz77',
  'rust',
  version : '2.1.0',
  meson_version : '>= 1.6.0',
  default_options : ['rust_std=2021'],
  license : 'MIT'
)
rust = import('rust')
features = {}
core2_options = {}
hashbrown_options = {}
rle_decode_fast_options = {}
required_deps = {}
if get_option('feature-default')
  required_deps += {'core2' : true, 'libflate' : true}
  features += {'std' : true, 'default' : true}
  core2_options += {'feature-std' : true}
  libflate_options += {'feature-std' : true}
endif
if get_option('feature-std')
  required_deps += {'core2' : true, 'libflate' : true}
  features += {'std' : true}
  core2_options += {'feature-std' : true}
  libflate_options += {'feature-std' : true}
endif
```

**To Reproduce**
meson.build:
```
project(
    'test',
    'rust',
    default_options: {'rust_std': '2021'},
    meson_version: '1.6.0',
)

matfile_proj = subproject('matfile')
```

matfile.wrap:
```
[wrap-file]
method = cargo
directory = matfile-0.5.0

source_url = https://crates.io/api/v1/crates/matfile/0.5.0/download
source_filename = matfile-0.5.0.tar.gz
source_hash = b22724209845c600f095bca5be902d636b4fc4f0b27fe5bb5d8314b5c117d0d6

[provide]
dependency_names = matfile
```

**Expected behavior**
The generated meson.build for libflate_lz77 does not reference an undefined variable.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? Native build
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.) Gentoo Linux
* what Python version are you using? 3.12.7
* what `meson --version` 1.6.0 (with patch from 5769074890e0a7f77ce1e607c8c680580dbb8b6f)
* what `ninja --version` if it's a Ninja build? 1.12.1

**Comment by peterdelevoryas** [comment]
I also have this issue, with a different project and dependency [here](https://github.com/peterdelevoryas/dbus-sensors/commit/bffbb42b4d63374817f40f2b9dd1ec01e87f708e). It's kind of blocking me from using Meson to build Rust, because there doesn't seem to be an option disable default features, like `--no-default-features` would? Based on [this comment](https://github.com/mesonbuild/meson/blob/master/mesonbuild/cargo/interpreter.py#L486-L487).

**Comment by eli-schwartz** [comment]
Poking around a bit in the libflate example, I notice that libflate depends on libflate_lz77.

libflate itself has these features:
```ini
[features]
default = ["std"]
std = ["libflate_lz77/std", "core2/std"]
```

Meanwhile, libflate_lz77 has these features:
```ini
[features]
default = ["std"]
std = ["core2/std", "libflate/std"]
```

It looks like... each one is trying to enforce on the other, to build with a consistent set of features? And lz77 wants to force the feature active for its own *parent* crate, not a dependency. Which is apparently legal. Is this only the case because libflate via path ../ is a dev-dependency, or is this generally legal?

I guess meson needs to handle arbitrary features getting set on things that aren't in dependencies.

**Comment by peterdelevoryas** [comment]
Yeah I suspect `libflate_lz77` has a `dev-dependency` on `libflate` because it's using it in doc-comment examples, which actually get compiled for some tests.

https://github.com/sile/libflate/blob/6676d36d381eec60f376337e528d658b16c87c7a/libflate_lz77/src/lib.rs#L117-L118

I think the issue is this code in meson is assuming that a dependency should be added whenever a crate tries to enable a feature on a dependency crate using `self_feature = ["_dep/dep_feature"]` syntax:

https://github.com/mesonbuild/meson/blob/master/mesonbuild/cargo/interpreter.py#L581-L582

I haven't quite gotten the fix right though, so, maybe I'm mistaken somehow.

**Comment by xclaesse** [comment]
First step at fixing this: https://github.com/mesonbuild/meson/pull/13976.

Configure passes but it has a bunch of build time errors, like:
```
FAILED: subprojects/zerocopy-0.7.32/libzerocopy.rlib 
rustc -C linker=cc --color=always -C debug-assertions=yes -C overflow-checks=no --crate-type rlib --edition=2018 -g --crate-name zerocopy --emit dep-info=subprojects/zerocopy-0.7.32/zerocopy.d --emit link=subprojects/zerocopy-0.7.32/libzerocopy.rlib --out-dir subprojects/zerocopy-0.7.32/libzerocopy.rlib.p -C metadata=adcab9b@@zerocopy@sta --cfg 'feature="simd"' ../subprojects/zerocopy-0.7.32/src/lib.rs
error: environment variable `CARGO_PKG_VERSION` not defined at compile time
    --> ../subprojects/zerocopy-0.7.32/src/lib.rs:1356:58
     |
1356 |     doc = concat!("[derive]: https://docs.rs/zerocopy/", env!("CARGO_PKG_VERSION"), "/zerocopy/derive.FromZeroes.html"),
     |                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = help: Cargo sets build script variables at run time. Use `std::env::var("CARGO_PKG_VERSION")` instead
     = note: this error originates in the macro `env` (in Nightly builds, run with -Z macro-backtrace for more info)
```

That requires using the `rustc --set-env` CLI. I think it got merged upstream but I'm not sure it's stable yet.

**Comment by xclaesse** [comment]
Seems rustc decided to close the issue.... someone will have to push them harder... https://github.com/rust-lang/rust/pull/119926

**Comment by bonzini** [comment]
Closing as the `options` variable are not generated anymore.
